Various OLM/OSDK 4.6 relnotes (#25624)
openshift/openshift-docs: release_notes/ocp-4-6-release-notes.adoc

[id="ocp-4-6-operators"]
=== Operators

[id="ocp-4-6-version-dependency"]
==== Operator version dependency

Operator developers can now ensure their Operators include dependencies on
specific versions of other Operators.

See xref:../operators/understanding/olm/olm-understanding-dependency-resolution.adoc#olm-understanding-dependency-resolution[Operator Lifecycle Manager dependency resolution] for more information.

[id="ocp-4-6-addtl-objects-bundle"]
==== Additional optional objects supported in Operator bundles

The Operator Bundle Format now supports the following optional Kubernetes objects:

* PodDisruptionBudget
** `policy/v1beta1`
* PriorityClass
** `scheduling.k8s.io/v1`
** `scheduling.k8s.io/v1beta1`
** `scheduling.k8s.io/v1alpha`
* VerticalPodAutoScaler

See xref:../operators/understanding/olm-packaging-format.adoc#olm-bundle-format-manifests-optional_olm-packaging-format[Operator Framework packaging formats] for more information.

[id="ocp-4-6-selective-mirroring-opm"]
==== Selective bundle image mirroring with `opm`

> Using the `opm` CLI, Operator administrators can now to select which bundle

Reviewer: ```suggestion
When using the `opm` CLI, Operator administrators can now to select which bundle
```

images to mirror.

See xref:../operators/admin/olm-restricted-networks.adoc#olm-restricted-networks[Pruning an index image] for more information.

[id="ocp-4-6-conversion-webhook-support"]
==== Conversion webhook support for global Operators

> Operator developers can now leverage conversion webhooks for Operators that

Reviewer: ```suggestion
Operator developers can now use conversion webhooks for Operators that
```

{product-title} 4.6 introduces the following notable technical changes.

[discrete]
[id="ocp-4-6-operator-catalogs-per-version"]
==== Default Operator catalogs now shipped per cluster version

> Starting in {product-title} 4.6 and later, the Red Hat-provided default catalogs used by Operator Lifecycle Manager and OperatorHub are now shipped as index images specific to the minor version of {product-title}. This allows Operator providers to ship intentional ranges of Operator versions per cluster version.

Reviewer: ```suggestion
Starting in {product-title} 4.6, the Red Hat-provided default catalogs used by Operator Lifecycle Manager and OperatorHub are now shipped as index images specific to the minor version of {product-title}. This allows Operator providers to ship intentional ranges of Operator versions per cluster version.
```

These index images, based on the Bundle Format, replace the App Registry catalog images, based on the deprecated Package Manifest Format, that are distributed for previous versions of {product-title} 4. {product-title} 4.1 through 4.5 will continue to share a single App Registry catalog.

[NOTE]
====
While App Registry catalog images are not distributed by Red Hat for {product-title} 4.6 and later, custom catalog images based on the Package Manifest Format are still supported.
====

[discrete]
[id="ocp-4-6-operator-catalogs-upgrade"]
===== Important Operator upgrade requirements

> Cluster administrators must ensure any Operators on their cluster installed using Operator Lifecycle Manager (OLM) have been updated to their latest channel and available version before upgrading to {product-title} 4.6. This ensures there is a valid upgrade path for the Operators as they switch from using the App Registry catalogs to the new index image-based catalogs.

Reviewer: ```suggestion
Cluster administrators must ensure any Operators on their cluster that are installed by using Operator Lifecycle Manager (OLM) have been updated to their latest channels and available version before upgrading to {product-title} 4.6. Updating these Operators ensures that there is a valid upgrade path for the Operators as they switch from using the App Registry catalogs to the new index image-based catalogs.
```

This is new functionality included with Kubernetes 1.19.

[discrete]
[id="ocp-4-6-operator-sdk-v-0-19-4"]
==== Operator SDK v0.19.4

{product-title} supports Operator SDK v0.19.4, which introduces the following
notable technical changes:

* Operator SDK is now aligned with the {product-title}-wide switch to using UBI-8
and Python 3. Downstream base images now use UBI-8 and include Python 3.
* The command `run --local` has been deprecated in favor of `run local`.
* The commands `run --olm` and `--kubeconfig` have been deprecated in favor of `run packagemanifests`.

> * The default CRD version  has been changed from `apiextensions.k8s.io/v1beta1` to `apiextensions.k8s.io/v1` for commands that create or generate CRDs.

Reviewer: ```suggestion
* The default CRD version has changed from `apiextensions.k8s.io/v1beta1` to `apiextensions.k8s.io/v1` for commands that create or generate CRDs.
```

Reviewer: I think you can use simpler tenses throughout this list.

* The `--kubeconfig` flag has been added to the `<run|cleanup> packagemanifests` command.
* The `scorecard` command has been removed and replaced by the Technology Preview `alpha scorecard` command, which is an enhanced implementation. See xref:../operators/operator_sdk/osdk-scorecard.adoc#osdk-scorecard[Validating Operators using the scorecard] for more information.

Ansible-based Operator enhancements include:

* The Ansible Operator is now available as a supported release.
* The Ansible Operator now includes a `healthz` endpoint and `liveness` probe.

Helm-based Operator enhancements include:

* Helm Operators can watch and reconcile when cluster-scoped release resources are changed.

> * Helm Operators can now reconcile logic using three-way strategic merge patches for native Kubernetes objects so that array patch strategies are correctly honored and applied.

Reviewer: ```suggestion
* Helm Operators can now reconcile logic by using three-way strategic merge patches for native Kubernetes objects so that array patch strategies are correctly honored and applied.
```
